src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContextHelpers.cs

```diff
@@ -94,6 +96,12 @@ private static void VerifyAction<TContext>(Action<TContext> action)
             {
                 throw new ArgumentNullException(nameof(action));
             }
+
+            // Disallow async methods to be registered.
+            if (action.GetMethodInfo().IsDefined(typeof(AsyncStateMachineAttribute)))
```

Comment:
Actually, I was wondering if there is a better way to do this, and the attribute [description](https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.asyncstatemachineattribute(v=vs.110).aspx) seems to suggest this is supported: `Indicates whether a method is marked with either the Async (Visual Basic) or async (C# Reference) modifier.` Adding @gafter for suggestion.